### PR TITLE
Configuration profiles

### PR DESCRIPTION
**Pull request by kasperski95**
--profile PROFILE, -p PROFILE

  Specifies active profile configuration.
  #### CI configuration
  ```toml title="protostar.toml"
  [protostar.shared_command_configs.ci]
  no_color=true
  ```
  `protostar -p ci test`
  
  #### Deployment configuration
  ```toml title="protostar.toml"
  [protostar.deploy.devnet]
  gateway_url="http://127.0.0.1:5050/"
  ```
  `protostar -p devnet deploy ...`

## EDIT
changed format to:
`[profile.PROFILE_NAME.protostar.COMMAND_NAME]`
- [x] add e2e test
- [x] update docs
- [x] update `--profile` description that it can't be configured from `protostar.toml`

### CODE REVIEW

**Comment by MaksymilianDemitraszek** [comment]
As we discussed please update to 
```
[ci.protostar.shared_command_configs]
```

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
LGTM, but I think we have no e2e test checking if config data is correctly used, can we add one? We can use profiles in this test to check if they work correctly. Integration test instead e2e will be fine either.

**Comment by kasperski95** [comment]
> As we discussed please update to
> 
> ```
> [ci.protostar.shared_command_configs]
> ```

without `profiles`?

**Comment by MaksymilianDemitraszek** [comment]
Im okay with both

File: website/docs/tutorials/03-project-initialization.md [hunk 1]
```diff
@@ -41,19 +41,35 @@ main = [
 ]
 
 ```
-Protostar's commands can be configured in the `protostar.toml`. This is number of arguments you would have to provide in the CLI otherwise. Protostar expects a section `["protostar.COMMAND_NAME"]` and a argument name with underscores (`_`) in place of dashes (`-`), for example:
+### Dynamic configuration
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/03-project-initialization.md:44]: I think Static and Dynamic are not very descriptive names and make no much sense in this context. What is dynamic about this argument? Are they changed by protostar dynamically?
Maybe `Project configuration` and `command arguments usign config` or something similar?

File: website/docs/tutorials/03-project-initialization.md [hunk 2]
```diff
@@ -41,19 +41,35 @@ main = [
...
-```
\ No newline at end of file
+```
+
+### Configuration profiles
+Configuration profiles provide a way to easily switch between Protostar configurations. This is especially useful for configuring StarkNet networks. Profiles inherit values from non-profiled configuration. In order to create a configuration profile, add a new section in `protostar.toml` with the following naming convention:<br/>  `["profile.PROFILE_NAME.protostar.COMMAND_NAME"]`.
+
+```toml title="protostar.toml"
+# ...
+["profile.ci.protostar.shared_command_configs"]
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/03-project-initialization.md:68]: This is a really cool example, gives a lot of intuition

File: protostar/start.py [hunk 3]
```diff
@@ -1,23 +1,28 @@
...
         protostar_cli,
-        default_value_provider=ArgumentValueFromConfigProvider(protostar_cli.project),
+        default_value_provider=ArgumentValueFromConfigProvider(
+            protostar_cli.project,
+            configuration_profile_name=ArgumentParserFacade(
+                ConfigurationProfileCLI(), disable_help=True
+            )
+            .parse(ignore_unrecognized=True)
+            .profile,
+        ),
```
Comment (MaksymilianDemitraszek) [inline protostar/start.py:21]:
```suggestion
        current_profile = ArgumentParserFacade(
                ConfigurationProfileCLI(), disable_help=True
            )
            .parse(ignore_unrecognized=True)
            .profile,

        default_value_provider=ArgumentValueFromConfigProvider(
            protostar_cli.project,
            configuration_profile_name=current_profile
        ),
```
Nitpick but I would prefer it this way, it's a bit messy right now

File: protostar/protostar_cli.py [hunk 4]
```diff
@@ -24,6 +24,38 @@
...
+            "[protostar.deploy.devnet]",
+            'gateway_url="http://127.0.0.1:5050/"',
+            "```",
+            "`protostar -p devnet deploy ...`" "",
+        ]
+    ),
+)
+
+
+class ConfigurationProfileCLI(CLIApp):
```
Comment (MaksymilianDemitraszek) [inline protostar/protostar_cli.py:52]: CLI is probably a bad name for this abstraction, right now, but I don't have a better idea

**Review by MaksymilianDemitraszek — APPROVED** [review]
LGTM but please refine section naming in docs, it's very misleading

File: protostar/protostar_cli.py (hunk 4 above)
Comment (MaksymilianDemitraszek) [reply protostar/protostar_cli.py:52]: Maybe we should split responsibilities here in the future somewhat

File: website/docs/tutorials/03-project-initialization.md (hunk 1 above)
Comment (kasperski95) [reply website/docs/tutorials/03-project-initialization.md:44]: Yep, not the best choice of words. It's dynamic from a Protostar developer perspective.

File: protostar/protostar_cli.py (hunk 4 above)
Comment (kasperski95) [reply protostar/protostar_cli.py:52]: We could, but it's not important right now. It doesn't add any value to the end-user and we have worse code in the repo.

Comment (MaksymilianDemitraszek) [reply protostar/protostar_cli.py:52]: Agreed

**Pull request merged**